sycl/include/sycl/ext/oneapi/bfloat16.hpp

```diff
@@ -156,12 +156,15 @@ class bfloat16 {
     return lhs = f;                                                            \
   }                                                                            \
   template <typename T>                                                        \
-  friend bfloat16 &operator op(bfloat16 & lhs, const T & rhs) {                \
+  friend std::enable_if_t<std::is_convertible_v<T, float>, bfloat16>           \
+      &operator op(bfloat16 & lhs, const T & rhs) {                            \
     float f = static_cast<float>(lhs);                                         \
     f op static_cast<float>(rhs);                                              \
     return lhs = f;                                                            \
   }                                                                            \
-  template <typename T> friend T &operator op(T & lhs, const bfloat16 & rhs) { \
+  template <typename T>                                                        \
+  friend std::enable_if_t<std::is_convertible_v<T, float>, T> &operator op(    \
```

Comment:
The current version of the extension returns `bfloat16` for both.

Reply:
This function cannot return 'bfloat16', it returns the modified `lhs`, which is of `T` type, e.g.: `T lhs; lhs += bfloat_var;'

Reply:
\+ @gmlueck

Reply:
@aelovikov-intel - this patch does not resolve the issue mentioned in this mini-discussion. It leaves the result type as it was before this patch as it is focused on different aspect.

The return type indeed should be fixed for operator+, not for operator+= where `lhs` is of `T` type, but that is a topic for a separate discussion & pull-request.

Reply:
There is no guarantee `lhs = f;` would work, is it?

Reply:
@aelovikov-intel I don't understand, can you elaborate your comment? What do you suggest to do in this mini-discussion?

Reply:
I don't think we should be working on this patch before the spec is fixed. If @gmlueck is on vacation, then the PR author should work with his backup to get the spec fixed first.

Reply:
We have already agreed with @gmlueck that std::is_convertible is what is meant in the existing SPEC.

Let's not wait until the ALL issues in spec resolved, it will take long and ESIMD will stay blocked by the wrong implementation.

Reply:
@steffenlarsen - can you please also review the patch as @gmlueck backup (as written in Greg's OOO).

Reply:
I don't see these operators in the [sycl_ext_oneapi_bfloat16 extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_oneapi_bfloat16.asciidoc). I only see the one with two bfloat16 operands. As such, I don't think they should be here unless we add them to the extension. Since they were here already and the extension is considered supported, we cannot just remove them, so we will have to consider deprecating them (or adding them to the extension.)

Reply:
@steffenlarsen - the spec has such lines:

> template <typename T>
>   friend bfloat16 operatorOP(const T &lhs, const bfloat16 &rhs) { /* …​ */ }
> 
> Construct a new instance of the bfloat16 class with the value of the new bfloat16 instance being the result of an OP arithmetic operation between the lhs of template type T and rhs bfloat16 value. Type T must be convertible to float.
> 
> OP is +, -, *, /

This patch here only syncs the implementation with this SPEC statement: "Type T must be convertible to float."

The attempt to 'static_cast<float>(lhs), where lhs is not convertible to float is simply a bug in implementation, fixed here.

Reply:
I agree with you, the changes in the patch adheres to the convertibility requirement in the extension, even though the operators here are technically not part of the extension. My primary point above is in response to the initial comment of

> The current version of the extension returns bfloat16 for both.

which for the operators right here aren't applicable as they are not even in the spec. The non-assignment binary operators below seem to have the correct types in accordance with the spec.

Reply:
I don't want to approve a PR that does not lessen the amount of technical debt in the codebase just because doing the right thing "will take long or even forever".

I can only assume that "forever" would happen if everybody would try to hack their way around (like this PR), instead of just doing the work on the specification.

It's not even the core SYCL, it's the extension that is part of this repository...

To clarify, I'm not going to prevent this merged if you find somebody else who'll give the approval, but it won't be me.

Reply:
Please see https://github.com/intel/llvm/pull/12477#issuecomment-1918713047. I agree there are things that need fixing in the specification, but these changes are in accordance with the current specification and are as such a bug-fix. Addressing additional problems in the extension cannot be made to the implementation immediately, as this isn't an experimental extension.

These operators are in a gray area, as they are not part of the extension, but since we made the mistake of exposing them, we should give them the same treatment we did with the extension interfaces, lest we deprecate them.

Reply:
> These operators are in a gray area, as they are not part of the extension, but since we made the mistake of exposing them, we should give them the same treatment we did with the extension interfaces, or deprecate them.

Then

1) It makes all the sense to deprecate them here
2) SFINAE enabler is wrong for those gray operators - `lhs = f` is not guaranteed to work. And the argument that the spec requires different requirements doesn't apply here (because I could use the same argument to remove them immediately).
3) I think it's debatable if a bugfix like this is even considered breaking change.